# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd
import os

In [2]:
pth1 = 'C:/pr/pr_1/data.csv' 
pth2 = '/datasets/data.csv'

if os.path.exists(pth1):
    data = pd.read_csv(pth1)
elif os.path.exists(pth2):
    data = pd.read_csv(pth2)
else:
    print('Something is wrong')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [3]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [5]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [6]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [7]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [8]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [9]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5], dtype=int64)

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [10]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [11]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5], dtype=int64)

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [12]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [13]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [14]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [15]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [16]:
data.duplicated().sum()

71

In [17]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [18]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [19]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [20]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [21]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [22]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [23]:
table = [] # Создаём пустой список, чтобы записать в него результат функции
def piv_table(a): # Создаём функцию
    table = data.pivot_table(index=a, values='debt', aggfunc=('mean', 'count', 'sum')) # Создаём сводную таблицу
    table.columns = ['Количество', 'Процент должников', 'Количество должников'] # Переименовываем столбцы таблицы
    table['Процент должников'] = table['Процент должников'].map('{:.2%}'.format) # Меняем формат вывода из десятичной дроби на проценты
    table = table.reset_index() # Делаем таблицу более удобочитаемой
    return table

In [24]:
piv_table('children') # Применяем функцию столбцу с данными о количестве детей

children  Количество Процент должников  Количество должников
0         0       14091             7.54%                  1063
1         1        4808             9.23%                   444
2         2        2052             9.45%                   194
3         3         330             8.18%                    27
4         4          41             9.76%                     4
5         5           9             0.00%                     0

In [25]:
def categorize_children(row):
    if row == 0:
        return '0'
    elif row == 1:
        return '1'
    elif row == 2:
        return '2'
    else: 
        return '3+'
data['children_category'] = data['children'].apply(categorize_children)
piv_table('children_category') 

children_category  Количество Процент должников  Количество должников
0                 0       14091             7.54%                  1063
1                 1        4808             9.23%                   444
2                 2        2052             9.45%                   194
3                3+         380             8.16%                    31

**Вывод:** По обобщённым данным видно, что наименьший процент задолженностей у клиентов без детей, а наибольший — у клиентов с двумя детьми. Также стоит отметить, что семьи с тремя и более детьми представлены относительно малой выборкой, на основе которой сложно сделать какие-либо выводы, так что не стоит исключать из потенциальной группы риска и их.

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [26]:
piv_table('family_status')

family_status  Количество Процент должников  Количество должников
0  Не женат / не замужем        2796             9.76%                   273
1              в разводе        1189             7.06%                    84
2         вдовец / вдова         951             6.62%                    63
3       гражданский брак        4134             9.31%                   385
4        женат / замужем       12261             7.56%                   927

**Вывод:** Наиболее ответственными в данном рассмотрении можно назвать категорию вдовцов (вдов). Следом за ними следуют клиенты в разводе и женатые/замужние. Наименее же ответственными являются неженатые (незамужние) и находящиеся в гражданском браке клиенты.

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [27]:
piv_table('total_income_category')

total_income_category  Количество Процент должников  Количество должников
0                     A          25             8.00%                     2
1                     B        5014             7.06%                   354
2                     C       15921             8.50%                  1353
3                     D         349             6.02%                    21
4                     E          22             9.09%                     2

**Вывод:** В силу малой репрезентативности категорий А, D и Е (ввиду малой выборки) можно заключить, что наиболее платёжеспособны клиенты с доходом от 200 тысяч до миллиона (категория В), а наименее — клиенты с доходом 50-200 тысяч (категория C).

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [28]:
piv_table('purpose_category')

purpose_category  Количество Процент должников  \
0    операции с автомобилем        4279             9.35%   
1  операции с недвижимостью       10751             7.26%   
2     получение образования        3988             9.25%   
3        проведение свадьбы        2313             7.91%   

   Количество должников  
0                   400  
1                   780  
2                   369  
3                   183

**Вывод:** Наименьшее количество задолженностей можно наблюдать по кредитам на операции с недвижимостью (вероятно, это обосновано страхом клиентов потерять место жительства). Наибольшее же количество должников в категории "операции с автомобилем".

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* Пропуски наблюдаются в столбцах с данными о трудовом стаже и заработке клиентов. Вероятно, клиенты, чьи данные пропущены, не были держателями зарплатных карт банка и по тем или иным причинам не предоставили соответствующие сведения в банк, что и привело к пропускам. Также, причина пропуска может быть чисто технологической — сбой в работе программы при выгрузке данных или их переносе из одной системы в другую. 

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* Использование медианного значения в данной ситуации обосновано тем, что оно менее чувствительно к "выбросам" в данных и отображает реальную картину лучше, чем среднее арифметическое

### Шаг 4: общий вывод.

Для построения модели кредитного скоринга были изучены взаимосвязи между платёжеспособностью клиентов банка и количеством их детей, семейным положением, уровнем дохода, а также целями, на которые клиентами был взят кредит. Вот основные итоги:

Количество детей          | Процент должников
--------------------------- | -------------
0 — 1  | 7.54 — 9.23 %
2 | 9.45 %

Семейное положение          | Процент должников
--------------------------- | -------------
Состояли (-ят) в браке  | 6.62 — 7.56 %
Не состояли (-ят) в браке         | 9.31 — 9.76 %

Уровень дохода  | Процент должников
------------- | -------------
50 — 200 тыс  | 7.06 %
200 тыс — 1 млн  | 8.50 %

Цель кредита  | Процент должников
------------- | -------------
Недвижимость, свадьба | 7.26 — 7.91 %
Образование, автомобиль | 9.25 — 9.35 %

Подробнее текстом:
- Наименьший процент должников у клиентов без детей: 7.54%. Наибольший — у клиентов с двумя детьми: 9.45%. 
- Лучше выплачивают кредиты клиенты состоящие (или состоявшие в браке). Процент должников среди них колеблется в пределах 6.62-7.56%. А вот среди людей, в официальном браке не состоявших, процент должников достигает 9.76%. 
- Клиенты с доходом 50-200 тысяч более ответственны. Процент должников среди них — 7.06%. А вот клиенты с доходом от 200 тысяч до миллиона — менее ответственны: среди них должников 8,5%.
- Меньше всего задолженностей по кредитам на недвижимость — 7.26%. Больше всего — по автомобилям: 9.35%.

Пара слов о том, как проводилось исследование:
- Прочитал файл с исходными данными, изучил общую информацию по этому файлу. Обнаружил пропуски в столбцах с данными о трудовом стаже и размере зарплаты клиентов. Пропуски заполнил медианным значением, так как оно менее чувствительно к выбросам в данных.
- Аномальные значения в столбце с данными о количестве детей было решено удалить, а в столбце с трудовым стажем — оставить неизменными, так как в дальнейшем анализе эти данные нам всё равно не пригодились.
- Обработал дубликаты в столбце с данными об образовании, удалил лишние строки.
- Категоризировал данные о доходах и целях взятия кредита, на основе которых впоследствии сделал описанные выше выводы.